File: WoofAndroid/presentation/grid/src/main/kotlin/com/perrystreet/woof/presentation/grid/ui/GridScreen.kt
package com.perrystreet.woof.presentation.grid.ui

import androidx.compose.foundation.lazy.grid.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.PreviewParameter
import com.perrystreet.woof.designsystem.atomic.molecules.placeholder.MolCardPlaceholder
import com.perrystreet.woof.designsystem.atomic.organisms.emptystate.OrgErrorState
import com.perrystreet.woof.designsystem.atomic.organisms.header.OrgNavigationHeaderBranded
import com.perrystreet.woof.designsystem.atomic.templates.TemplateCenteredContent
import com.perrystreet.woof.designsystem.atomic.templates.TemplateGrid
import com.perrystreet.woof.designsystem.preview.PreviewDevices
import com.perrystreet.woof.designsystem.preview.ThemeProvider
import com.perrystreet.woof.designsystem.preview.ThemedScreenPreview
import com.perrystreet.woof.designsystem.theme.ITheme
import com.perrystreet.woof.presentation.grid.ui.components.DogCell
import com.perrystreet.woof.presentation.grid.ui.preview.GridPreviewData
import com.perrystreet.woof.presentation.grid.uimodel.DogCellUIModel
import com.perrystreet.woof.presentation.grid.viewmodel.GridViewModel
import com.perrystreet.woof.resources.R

@Composable
fun GridScreen(
    state: GridViewModel.State,
    onCellAppear: (DogCellUIModel) -> Unit,
    onCellTap: (DogCellUIModel) -> Unit,
    onRetryTap: () -> Unit,
) {
    when (state) {
        GridViewModel.State.Loading -> GridLoadingScreen()
        is GridViewModel.State.Loaded -> GridLoadedScreen(
            state = state,
            onCellAppear = onCellAppear,
            onCellTap = onCellTap,
        )
        GridViewModel.State.Error -> GridErrorScreen(onRetryTap = onRetryTap)
    }
}

@Composable
private fun GridLoadingScreen() {
    TemplateGrid(topBar = { OrgNavigationHeaderBranded() }) {
        items(PlaceholderCellCount) {
            MolCardPlaceholder()
        }
    }
}

@Composable
private fun GridLoadedScreen(
    state: GridViewModel.State.Loaded,
    onCellAppear: (DogCellUIModel) -> Unit,
    onCellTap: (DogCellUIModel) -> Unit,
) {
    TemplateGrid(topBar = { OrgNavigationHeaderBranded() }) {
        items(items = state.cells, key = { cell -> cell.id }) { cell ->
            DogCell(cell = cell, onCellAppear = onCellAppear, onCellTap = onCellTap)
        }
        items(state.loadingMoreCellCount) {
            MolCardPlaceholder()
        }
    }
}

@Composable
private fun GridErrorScreen(onRetryTap: () -> Unit) {
    TemplateCenteredContent(topBar = { OrgNavigationHeaderBranded() }) {
        OrgErrorState(
            title = stringResource(R.string.grid_error_title),
            message = stringResource(R.string.grid_error_message),
            actionText = stringResource(R.string.grid_error_retry),
            onActionTap = onRetryTap,
        )
    }
}

private const val PlaceholderCellCount = 30

@PreviewDevices
@Composable
private fun GridScreenLoadedPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        GridScreen(
            state = GridPreviewData.loaded(),
            onCellAppear = {},
            onCellTap = {},
            onRetryTap = {},
        )
    }
}

@PreviewDevices
@Composable
private fun GridScreenLoadingPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        GridScreen(
            state = GridViewModel.State.Loading,
            onCellAppear = {},
            onCellTap = {},
            onRetryTap = {},
        )
    }
}

@PreviewDevices
@Composable
private fun GridScreenErrorPreview(@PreviewParameter(ThemeProvider::class) theme: ITheme) {
    ThemedScreenPreview(theme = theme) {
        GridScreen(
            state = GridViewModel.State.Error,
            onCellAppear = {},
            onCellTap = {},
            onRetryTap = {},
        )
    }
}

File: WoofIOS/SwiftPackages/Presentation/Grid/Sources/PresentationGrid/UI/GridScreen.swift
import DesignSystem
import Resources
import SwiftUI

public struct GridScreen: View {
    private let state: GridViewModel.State
    private let onCellAppear: (DogCellUIModel) -> Void
    private let onCellTap: (DogCellUIModel) -> Void
    private let onRetryTap: () -> Void

    public init(
        state: GridViewModel.State,
        onCellAppear: @escaping (DogCellUIModel) -> Void,
        onCellTap: @escaping (DogCellUIModel) -> Void,
        onRetryTap: @escaping () -> Void
    ) {
        self.state = state
        self.onCellAppear = onCellAppear
        self.onCellTap = onCellTap
        self.onRetryTap = onRetryTap
    }

    public var body: some View {
        switch state {
        case .loading:
            GridLoadingScreen()
        case .loaded(let loaded):
            GridLoadedScreen(state: loaded, onCellAppear: onCellAppear, onCellTap: onCellTap)
        case .error:
            GridErrorScreen(onRetryTap: onRetryTap)
        }
    }
}

private struct GridLoadingScreen: View {
    private static let placeholderCellCount = 30

    var body: some View {
        TemplateGrid(topBar: { OrgNavigationHeaderBranded() }) {
            ForEach(0..<Self.placeholderCellCount, id: \.self) { _ in
                MolCardPlaceholder()
            }
        }
    }
}

private struct GridLoadedScreen: View {
    let state: GridViewModel.State.Loaded
    let onCellAppear: (DogCellUIModel) -> Void
    let onCellTap: (DogCellUIModel) -> Void

    var body: some View {
        TemplateGrid(topBar: { OrgNavigationHeaderBranded() }) {
            ForEach(state.cells) { cell in
                DogCell(cell: cell, onCellAppear: onCellAppear, onCellTap: onCellTap)
            }
            ForEach(0..<state.loadingMoreCellCount, id: \.self) { _ in
                MolCardPlaceholder()
            }
        }
    }
}

private struct GridErrorScreen: View {
    let onRetryTap: () -> Void

    var body: some View {
        TemplateCenteredContent(topBar: { OrgNavigationHeaderBranded() }) {
            OrgErrorState(
                title: L10n.Grid.errorTitle,
                message: L10n.Grid.errorMessage,
                actionText: L10n.Grid.errorRetry,
                onActionTap: onRetryTap
            )
        }
    }
}

#Preview("Loaded") {
    ThemedScreenPreview(theme: WoofTheme.light()) {
        GridScreen(state: GridPreviewData.loaded(), onCellAppear: { _ in }, onCellTap: { _ in }, onRetryTap: {})
    }
}

#Preview("Loading") {
    ThemedScreenPreview(theme: WoofTheme.light()) {
        GridScreen(state: .loading, onCellAppear: { _ in }, onCellTap: { _ in }, onRetryTap: {})
    }
}

#Preview("Error") {
    ThemedScreenPreview(theme: WoofTheme.dark()) {
        GridScreen(state: .error, onCellAppear: { _ in }, onCellTap: { _ in }, onRetryTap: {})
    }
}
